### PR TITLE
feat(suite): auto-select a newly connected device when none is present

### LINKED ISSUE #30929: Automatically switch to a newly connected device when no other device is connected
When a new device is connected, the app could automatically switch to it if there is no other device currently connected or selected.

## Considerations

Please account for reconnection scenarios, especially:

- Firmware updates
- Bootloader mode transitions

As Szymon pointed out, a device may disconnect and reconnect during a firmware update. Any new auto-switching logic should avoid interfering with this flow or treating the reconnecting device as an unrelated newly connected device.

### PR DESCRIPTION
**Pull request by vojtatranta**
Switch to a newly connected device when the selected wallet has no device connected (e.g. a remembered one) and nothing else competes for the selection.

Skipped during a firmware installation, because the updated device disconnects and reconnects in bootloader mode without its `id`, and must not be taken for an unrelated newly connected device.

Closes #30929

## Notes for QA

To test out:
- have TS7 already paried
- pair another wired device
- turn off BT on TS7 
- select the wired device
- disconnect the Wired device
- turn ON BT on TS7
- observe that TS7 wallet is selected

**Please test that this process doesn't break the update FW process of some wired device** (do onboarding on a wirde device, and during that flow, connect the TS7 via BT as described above)

**Also please test that auto-selecting still works after a firmware update is over** — both after a successful one and after a cancelled/failed one:
- do a FW update on a wired device, then close the firmware modal
- disconnect the wired device
- turn ON BT on TS7 and observe that the TS7 wallet is selected
- repeat the same, but cancel the FW update instead of finishing it

Auto-selecting is deliberately suspended while a firmware update is in progress (the updated device reconnects in bootloader mode without its id and must not be taken for an unrelated device), and it is resumed when the firmware flow is closed. If the flow ever ends without being closed properly, auto-selecting would stay off until the app is restarted, which is what this scenario checks.

## Related Issue

Resolve #30929

## 🌐 Preview deployments

🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/30929-bt-device-auto-connect/web/](https://dev.suite.sldev.cz/suite-web/30929-bt-device-auto-connect/web/)

<details>
<summary><h3>🤖 LLM Test Recommendations</h3></summary>

**Summary:** The changes center on device selection logic (getShouldSelectConnectedDevice) and its use in selectors and the discovery thunk. Risk is concentrated in flows that switch, reconnect, eject, or discover wallets. The recommended high-priority tests target discovery, multi-session device acquisition, and wallet rediscovery. Medium-priority tests extend coverage to passphrase reconnection, remembered devices, device forgetting, and disconnected-state handling. The remaining changed files are test fixtures, unit tests, or an export index with no e2e coverage needed.

<details><summary><b>Changed files (7)</b></summary>

- `packages/suite/src/actions/suite/__fixtures__/suiteActions.ts`
- `packages/suite/src/actions/suite/suiteActions.test.ts`
- `suite-common/device/src/deviceSelectors.ts`
- `suite-common/device/src/getShouldSelectConnectedDevice.test.ts`
- `suite-common/device/src/getShouldSelectConnectedDevice.ts`
- `suite-common/device/src/index.ts`
- `suite-common/wallet-core/src/discovery/selectDeviceThunk.ts`

</details>

**Recommended tests (7)**

<details><summary>🔴 <b>High priority (3)</b></summary>

- `suite/e2e/tests/general/wallet-discovery.test.ts` — Directly exercises device ejection, adding a standard wallet, and discovery — all flows that depend on selecting the correct connected device via the changed device selection logic and discovery thunk.
- `suite/e2e/tests/suite/multiple-sessions.test.ts` — Tests Bridge session stealing/reclaiming across tabs and the 'Use Here' vs 'Connected' device status, which directly relies on device selection and acquisition logic touched by the changed selectors and thunk.
- `suite/e2e/tests/wallet/discovery.test.ts` — Activates many coins, triggers discovery, and verifies recovery after a page reload mid-discovery — a scenario where selectDeviceThunk and device selection decisions are heavily exercised.

</details>

<details><summary>🟡 <b>Medium priority (4)</b></summary>

- `suite/e2e/tests/settings/forget-device.test.ts` — Forgetting devices removes them from the device pool; regressions in device selection could cause stale selection state or incorrect onboarding redirects after a device is forgotten.
- `suite/e2e/tests/metadata/legacy/remembered-device.test.ts` — Covers remembered devices while the emulator is powered off and after reconnection; device selection must correctly fall back to and resume from a remembered wallet state.
- `suite/e2e/tests/passphrase/passphrase-reconnection.test.ts` — Adds a hidden wallet, disconnects/reconnects the device, and verifies cached passphrase wallets remain selectable — directly tests device selection after reconnect.
- `suite/e2e/tests/wallet/without-device.test.ts` — Powers off the device mid-send flow and checks disconnected state handling; the changed selection logic controls how Suite behaves when the previously selected device disappears.

</details>

⚠️ **Changes with no test coverage (4)**
- `packages/suite/src/actions/suite/__fixtures__/suiteActions.ts`
- `packages/suite/src/actions/suite/suiteActions.test.ts`
- `suite-common/device/src/getShouldSelectConnectedDevice.test.ts`
- `suite-common/device/src/index.ts`

_Updated: 2026-08-17T12:58:56.363Z_

</details>

## 🔍 Currents Test Results

🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=30929-bt-device-auto-connect&lookback=7)

🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=30929-bt-device-auto-connect&lookback=7)

🔍 **Suite native android test results:** [View in Currents](https://app.currents.dev/projects/iUe1Y4/runs?branch=30929-bt-device-auto-connect&lookback=7)

## 🔒 Quarantined E2E Tests

<details><summary>Trezor Suite (web) — 8 test(s)</summary>

| Test | Type |
|------|------|
| TrezorConnect popup web > call cancelled from calling application | 🤖 auto |
| Trading - DEX swap (LI.FI) > User can swap ETH to USDC via LI.FI DEX | 🤖 auto |
| Send - Solana > Send max | 🤖 auto |
| Passphrase > basic flow | 🤖 auto |
| Passphrase reconnection > after device is reconnected passphrase needs to be confirmed | 🤖 auto |
| Quarantine test: "Recovery - dry run,Recovery with device reconnection" | 🙋 manual |
| Quarantine test: "TrezorConnect webextension -> Suite Web,second call after popup was closed by user should work" | 🙋 manual |
| Quarantine test: "Recovery T2T1 - dry run,Recovery after partial recovery" | 🙋 manual |

</details>

_Updated: 2026-08-17T13:01:41.668Z • 8 test(s) total_

<details><summary>Trezor Suite (desktop) — 5 test(s)</summary>

| Test | Type |
|------|------|
| Passphrase reconnection > after device is reconnected passphrase needs to be confirmed | 🤖 auto |
| Trading - Swap > Swap SOL USDT token to ETH | 🤖 auto |
| Trading - Swap > Swap SOL to BTC | 🤖 auto |
| Quarantine test: "Recovery - dry run,Recovery after partial recovery" | 🙋 manual |
| Quarantine test: "Recovery - dry run,Recovery with device reconnection" | 🙋 manual |

</details>

_Updated: 2026-08-17T13:00:39.786Z • 5 test(s) total_

### CODE REVIEW

File: suite-common/wallet-core/src/discovery/selectDeviceThunk.ts [hunk 1]
```diff
@@ -52,9 +54,17 @@ export const selectDeviceThunk = createThunk<
...
     ({ device }, { dispatch, getState, rejectWithValue }) => {
-        if (!isNative() && !selectIsSameOrNewDevice(getState(), device)) {
+        const shouldSelect =
             // Select automatically when it is the first known device (none selected),
             // or when we connected physical device corresponding to a selected remembered wallet.
+            selectIsSameOrNewDevice(getState(), device) ||
+            // Or when the selected wallet has no device present and this is the only one connected.
+            // Never during a firmware installation - the updated device disconnects and reconnects
+            // in bootloader mode without its `id`, and must not be taken for an unrelated device.
+            (!selectIsFirmwareInstallationRunning(getState()) &&
```
Comment (vojtatranta) [inline suite-common/wallet-core/src/discovery/selectDeviceThunk.ts]: @vojtatranta review this

**Comment by vojtatranta** [comment]
I need to manually tests this to be sure. This sucks.

File: suite-common/wallet-core/src/device/deviceThunks.ts [hunk 2]
```diff
@@ -84,6 +86,14 @@ export const handleDeviceDisconnect = createThunk<
     if (!selectedDevice) return;
     if (selectedDevice.path !== device.path) return;
 
+    // The device being updated disconnects and reconnects on its own, in bootloader mode and then
```
Comment (vojtatranta) [inline suite-common/wallet-core/src/device/deviceThunks.ts]: I'll separte this to ad iffernt PR

File: suite-common/wallet-core/src/discovery/selectDeviceThunk.ts [hunk 3]
```diff
@@ -51,19 +55,26 @@ export const selectDeviceThunk = createThunk<
...
-            return rejectWithValue('no-need-to-select');
+        const { shouldSelect, reason } = getShouldSelectConnectedDevice({
+            connectedDevice: device,
+            selectedDevice: selectSelectedDevice(getState()),
+            physicalDeviceWallets: selectPhysicalDeviceWallets(getState()),
+            firmware: selectFirmware(getState()),
+        });
+
+        // Mobile has a single device at a time, so it always follows the one that connected.
+        if (!isNative() && !shouldSelect) {
```
Comment (Lemonexe) [inline suite-common/wallet-core/src/discovery/selectDeviceThunk.ts]: Tiny nit: this condition is hard to understand (even before this PR ofc). I  hope that making a native-only logical branch will be clearer:

```suggestion
        if(isNative()) {
            dispatch(selectDeviceThunk({ device }));

            return
        }

        if (!shouldSelect) {
```

File: suite-common/device/src/getShouldSelectConnectedDevice.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import { type Device } from '@trezor/connect';
+
+/** Why the connected device takes the selection over. */
+export type SelectDeviceReason =
+    /** No wallet is selected yet, the first device that shows up takes the selection. */
+    | 'no-selected-device'
+    /** The physical device of the selected wallet, remembered or freshly reconnected. */
+    | 'same-device'
+    /** The selected wallet has no device present and nothing else competes for the selection. */
+    | 'only-connected-device';
```
Comment (Lemonexe) [inline suite-common/device/src/getShouldSelectConnectedDevice.ts:11]: This is the new case that this PR aims to cover, is that right?

File: suite-common/device/src/getShouldSelectConnectedDevice.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    // A nullish `descriptor.id` means "unknown", not "no id": it is optional in the transport
+    // layer and old bridge does not report it at all. Two unknowns must never look like a match.
+    if (descriptorId != null && descriptorId === b?.descriptor?.id) {
+        return true;
+    }
+
+    // `descriptor.id` is scoped to the transport, so the same device reached over another one does
+    // not match by it. `id` is reported whenever the device is not in bootloader mode.
+    return a?.id != null && a.id === b?.id;
```
Comment (Lemonexe) [inline suite-common/device/src/getShouldSelectConnectedDevice.ts:70]: ahh, this is another new piece of logic that the previous code did not cover.
If I understand correctly, this is to select the THP unacquired device over different transport than remembered wallet (i.e. can be the same TS7 device, but the remembered one is acquired via cable, and the incoming one is unacquired via BT), is that right?

File: suite-common/device/src/getShouldSelectConnectedDevice.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ * wallet. Pure, so that the rules can be read and tested in one place - the caller only feeds it
+ * the state it needs.
+ */
+export const getShouldSelectConnectedDevice = ({
+    connectedDevice,
+    selectedDevice,
+    physicalDeviceWallets,
+    firmware,
+}: GetShouldSelectConnectedDeviceParams): DeviceSelectionDecision => {
+    if (connectedDevice === undefined) {
```
Comment (Lemonexe) [inline suite-common/device/src/getShouldSelectConnectedDevice.ts]:
```suggestion
    // This function should be called when there is an incoming device being connected, but just to cover all cases:
    if (connectedDevice === undefined) {
```

File: suite-common/device/src/getShouldSelectConnectedDevice.test.ts [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+);
+const DEVICE_B_UNACQUIRED = withDescriptor(
+    mockSuiteDevice({ path: '2', type: 'unacquired', connected: true }),
+    descriptorB,
+);
+
+const NO_FIRMWARE_UPDATE = { status: 'initial' } as const;
+
+describe('getShouldSelectConnectedDevice', () => {
+    it('selects nothing when no device is connected', () => {
```
Comment (Lemonexe) [inline suite-common/device/src/getShouldSelectConnectedDevice.test.ts]: Tiny nits, but the current wording _is_ misleading:

🤔 maybe the parameter could be renamed `connectedDevice` → `incomingDevice`, then all occurences would be self-explanatory.

```suggestion
    it('selects nothing when no device is being connected', () => {
```

File: suite-common/device/src/getShouldSelectConnectedDevice.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    connectedDevice: Device | TrezorDevice | undefined;
+    /** The wallet selected so far, whose device may or may not be present. */
+    selectedDevice: TrezorDevice | undefined;
+    /** All wallets of all physical devices, i.e. `selectPhysicalDeviceWallets`. */
+    physicalDeviceWallets: readonly TrezorDevice[];
+    firmware: FirmwareState;
+};
+
+/**
+ * `descriptor.id` identifies the physical device across reconnects, unlike `path`, which is
```
Comment (Lemonexe) [inline suite-common/device/src/getShouldSelectConnectedDevice.ts]: does it? Maybe yeah, I just don't know. Did you verify this in Connect/Transport?

**Review by Lemonexe — APPROVED** [review]
Have not tested but codewise LGTM.
This will definitely need a lot of creative QA to test for edge cases..

File: suite-common/device/src/getShouldSelectConnectedDevice.ts (hunk 4 above)
Comment (vojtatranta) [reply suite-common/device/src/getShouldSelectConnectedDevice.ts:11]: yes

File: suite-common/device/src/getShouldSelectConnectedDevice.ts (hunk 8 above)
Comment (vojtatranta) [reply suite-common/device/src/getShouldSelectConnectedDevice.ts]: will check

File: suite-common/device/src/getShouldSelectConnectedDevice.ts (hunk 5 above)
Comment (vojtatranta) [reply suite-common/device/src/getShouldSelectConnectedDevice.ts:70]: No, this is here only for case when firmware update is ongoing to detect that thte same device was connected that's not acquired.

File: suite-common/device/src/getShouldSelectConnectedDevice.ts (hunk 8 above)
Comment (vojtatranta) [reply suite-common/device/src/getShouldSelectConnectedDevice.ts]: descriptor.id does do it refer to the same device necesarily, but ot hte same transport

File: suite-common/device/src/getShouldSelectConnectedDevice.ts (hunk 5 above)
Comment (vojtatranta) [reply suite-common/device/src/getShouldSelectConnectedDevice.ts:70]: earlier, there was no guard during hte FW run, so therotically, if you connected a different deivce during the FW run it would break it. NOw it at least checks if it should select / not select a same device.

The reason for thsi gbeing here is to selected a device that's in firmware update again once it reconnected

**Comment by honzauher** [comment]
QA OK

Tested before merging, also tested now


FW update **cancel** scenario:

https://github.com/user-attachments/assets/10ff769a-2bf1-4a68-bb55-d91c29cba860

Autoselect scenario:

https://github.com/user-attachments/assets/da3752ed-9833-4fc9-b3dd-f2912e773888

**Info:**
 - Suite version: desktop 26.9.0 (64d40a90649289122cd27ad5ea4f7dd7b7949a0e)
 - Browser: Trezor Suite 26.9.0
 - OS: MacIntel
 - Screen: 1512x982
 - Device: Trezor T2B1 2.12.4 universal (revision 30be4e8c9488eeab68f994af23b3d9c9b7334266)
 - Transport: BridgeTransport 3.2.1, BluetoothTransport

**Comment by vojtatranta** [comment]
@honzauher got this report https://github.com/trezor/trezor-suite/issues/31911

So we should have tested this aswell, damn it

**Comment by honzauher** [comment]
> @honzauher got this report #31911
> 
> So we should have tested this aswell, damn it

Is that actually caused by this PR?

**Comment by vojtatranta** [comment]
> > @honzauher got this report #31911
> > So we should have tested this aswell, damn it
> 
> Is that actually caused by this PR?

I was told so, didnt' verify

**Comment by vojtatranta** [comment]
> > > @honzauher got this report #31911
> > > So we should have tested this aswell, damn it
> > 
> > 
> > Is that actually caused by this PR?
> 
> I was told so, didnt' verify

fixing PR https://github.com/trezor/trezor-suite/pull/31922

pls if you had a while, please check it, thank you https://github.com/trezor/trezor-suite/pull/31922
